README.md
# Portecle [![CI status](https://github.com/scop/portecle/workflows/CI/badge.svg)](https://github.com/scop/portecle/actions?query=workflow%3ACI) [![Download](https://img.shields.io/sourceforge/dt/portecle.svg)](https://sourceforge.net/projects/portecle/files/latest/download)

Portecle is a user friendly GUI application for creating, managing and
examining keystores, keys, certificates, certificate requests,
certificate revocation lists and more.

Currently, Portecle can be used to, for example:

- Create, load, save, and convert keystores.
- Generate DSA and RSA key pair entries with self-signed X.509
  certificates.
- Import X.509 certificate files as trusted certificates.
- Import key pairs from PKCS #12 files.
- Clone and change the password of key pair entries and keystores.
- View the details of certificates contained within keystore entries,
  certificate files, and SSL/TLS connections.
- Export keystore entries in a variety of formats.
- Generate and view certification requests (CSRs).
- Import Certificate Authority (CA) replies.
- Change the password of key pair entries and keystores.
- Delete, clone, and rename keystore entries.
- View the details of certificate revocation list (CRL) files.
- Generate SM2 key pair entries with self-signed X.509 certificates.
- Generate SM2 PKCS10 certificates request.
- Support SM2 Certificate Authority (CA) replies.

Getting up and running with Portecle is quick and easy.  Everything
you need to know is detailed below. Being written in Java, Portecle
will run on any machine that has a suitable Java runtime environment
installed. 

You can access the online help of Portecle from within the Portecle
GUI, or online at http://portecle.sourceforge.net/#docs

## Installing

First, you'll need a suitable Java runtime environment installed.
Java SE version 7 or later is required; see for example
https://www.oracle.com/technetwork/java/index.html and
https://openjdk.java.net/ for available versions and install
instructions.

Apart from Java, the default binary distribution of Portecle contains
everything you'll need to run it. The easiest way to install it is to
unzip the binary distribution to a directory somewhere on your
filesystem.

The binary distribution contains the Portecle jar file
(`portecle.jar`) as well as Bouncy Castle provider and PKIX jars
(`bcprov.jar`, `bcpkix.jar`) for Java SE 7.  If you wish to run
Portecle with a later Java version, or update the bundled Bouncy
Castle jars for some other reason, simply download an update for your
version of Java from https://www.bouncycastle.org/ and place the jars
into the same directory as `portecle.jar` with the names `bcprov.jar`
and `bcpkix.jar`, overwriting the existing one already there (if any).
The binary distribution also contains icons for use with Portecle.

Portecle can additionally use the GNU Classpath (version 0.90 or
later) security providers if they are installed. Support for GNU
Keyring (GKR) keystores requires these providers. For more
information about GNU Classpath, see
https://www.gnu.org/software/classpath/

Depending on your Portecle usage patterns, the Bouncy Castle provider
may require the JCE unlimited strength jurisdiction policy files
installed to function properly.  See "IMPORTANT NOTES" at
https://www.bouncycastle.org/documentation.html. Failures related to
lack of these policy files usually manifest themselves as errors
loading keystores with an error message like "Unsupported keysize or
algorithm parameters" or "Illegal key size" when trying to import
keys.

The default way of running Portecle uses the `java -jar` method, which
means that the `Class-Path` defined in `portecle.jar`'s `MANIFEST.MF`
will be used to locate all classes. You can also invoke Portecle by
its "main" class, `net.sf.portecle.FPortecle`. This method allows you
to use a Bouncy Castle provider jar elsewhere on your filesystem.

The following chapters contain examples how to run Portecle; all of
the examples assume that the JRE/JDK `bin` directory has been added to
your `PATH` environment variable.

### Windows Command Line

Assuming you have an appropriate JRE/JDK installed and have placed the
Portecle and Bouncy Castle provider JAR files into a directory
`c:\java` you can run Portecle like so:

```
java -jar c:\java\portecle.jar
```

In most setups, if `portecle.jar`, `bcprov.jar`, and `bcpkix.jar` were
installed as instructed above, Portecle can also be run by
double-clicking `portecle.jar` in the Windows Explorer.

If you wish to manage the jar locations yourself, use Java's `-cp`
option for that, and `net.sf.portecle.FPortecle` as the class to
launch.

### UNIX Command Line

Assuming you have an appropriate JRE/JDK installed and have placed the
Portecle and Bouncy Castle provider jar files into a directory
`/usr/share/java` you can run Portecle like so:

```
java -jar /usr/share/java/portecle.jar
```

If you wish to manage the jar locations yourself, use Java's `-cp`
option for that, and `net.sf.portecle.FPortecle` as the class to
launch.

### Experimental Features

Portecle releases may contain experimental features that are not
enabled by default.  These have known limitations or incomplete
implementations that make them unsuitable for production use, but they
may be valuable for early adopters. To enable these features, use
`-Dportecle.experimental=true` in your Portecle invocation command
line. See the file [NEWS.txt](NEWS.txt) for information about status
of current experimental features.

## Copyright and License

Copyright © 2004 Wayne Grant, 2004 Mark Majczyk, 2004-2019 Ville Skyttä

Portecle is free software; you can redistribute it and/or modify it
under the terms of the GNU General Public License as published by the
Free Software Foundation; either version 2 of the License, or (at your
option) any later version.

Portecle is distributed in the hope that it will be useful, but
WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
General Public License for more details.

You should have received a copy of the GNU General Public License
along with Portecle, see the file LICENSE.txt; if not, write to the
Free Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston,
MA 02110-1301 USA

## Contact

For contact information and issue tracking facilities,
see Portecle's project pages at:
- https://github.com/scop/portecle
- https://sourceforge.net/projects/portecle/


NEWS.txt
New in Portecle x.xx (201x-xx-xx):
----------------------------------


Enhancements:
- Support setting multiple and different types of Subject Alternative Names
  when generating certificates (#55).

Bug fixes:
- Work around Java 8+ possibly opening PKCS#12 keystores as JKS (#57).

Miscellaneous:
- Bundled Bouncy Castle jars upgraded to version 1.62.
- Document use of proxy servers, see doc/use-proxies.html (#44).

New in Portecle 1.11 (2017-12-12):
----------------------------------

Enhancements:
- Add support for BCFKS keystores (#48).
- Support setting Subject Alternative (DNS) Name when generating
  certificates (feature request #22).
- Preserve extensions when renewing self-signed certificates.
- Recognition for some additional certificate extension values.
- Link fixes and updates to use HTTPS in documentation and UI.
- Remove link to decommissioned project mailing lists.
- Update AppStream metadata and desktop entry files to current standards.
- New icons and OS X icon image support (Martin W. Kirst, #47).
- New alt[1-3] icon sets (Luya Tshimbalanga).

Bug fixes:
- Fix exception when opening the PKCS#11 provider selection dialog.
- Fix exception with nulls when rendering key store and extensions tables.

Miscellaneous:
- Generated X.509 certificates are now version 3 ones.
- Bundled Bouncy Castle jars upgraded to version 1.58; 1.56 is the new
  minimum required version.

New in Portecle 1.10 (2016-02-04):
----------------------------------

Enhancements:
- Support SNI in SSL/TLS connections.
- Recognition for some additional certificate extension values.

Miscellaneous:
- Bundled Bouncy Castle jars upgraded to version 1.54.

New in Portecle 1.9 (2015-06-23):
---------------------------------

Enhancements:
- Make SHA256withRSA the default for new RSA key pairs.
- Drop hardwired upper key size limit for generated DSA keys.
- Allow key pair generation using SHA384withDSA and SHA512withDSA.
- AppData XML and desktop entry updates.

Bug fixes:
- Fix MissingResourceException saving BKS-V1 keystores (bug #20).

Miscellaneous:
- Bundled Bouncy Castle jars upgraded to version 1.52; 1.51 is the
  new minimum required version.

New in Portecle 1.8 (2014-01-14):
---------------------------------

Enhancements:
- Add support for BKS-V1 keystores (thanks to Jan Peter Stotz, patch #3).
- Command line arg that looks like host:port opens SSL/TLS connection dialog.
- Finnish translation (incomplete).

Bug fixes:
- E-mail address validation no longer rejects certain addresses (bug #18).

Miscellaneous:
- Java SE 7 or later is now required.
- Bundled Bouncy Castle jars upgraded to version 1.50 which is also the
  new minimum required version.

New in Portecle 1.7 (2011-01-23):
---------------------------------

Enhancements:
- Certificate extension rendering improvements.
- Miscellaneous user interface improvements.
- Recognition for some additional certificate extension values.
- Documentation updates.

Bug fixes:
- Unknown key sizes no longer prevent viewing certificates or generating
  key store reports.

Miscellaneous:
- Portecle is now available through Java Web Start at
  http://portecle.sourceforge.net/webstart/portecle.jnlp
- Default for newly generated RSA keys is now 2048 bits.

New in Portecle 1.6 (2011-01-15):
---------------------------------

New features:
- Support for dragging and dropping files to the main window.

Enhancements:
- Certificate extension rendering improvements.
- Miscellaneous user interface improvements.
- Error logging improvements.
- Don't refuse to start with unrecognized JRE versions.
- Try to avoid file chooser delays on Windows (#3129497).

Bug fixes:
- Swing event dispatch thread related fixes (#3129497).

Miscellaneous:
- Bundled Bouncy Castle provider jar upgraded to version 1.45.

New in Portecle 1.5 (2010-01-03):
---------------------------------

New features:
- Recognition and decoding support for some additional certificate
  extensions.
- Support for DSA certificates using the SHA224withDSA, SHA256withDSA,
  SHA384withDSA, and SHA512withDSA signature algorithms.
- In addition to keystores, certificates, CRL's and CSR's can now be
  passed as command line arguments.

Enhancements:
- Improved import of trusted certificates from PEM files containing also
  password protected keys.
- Make Esc key close more dialog boxes.
- Certificate extension rendering improvements.
- Other user interface improvements.

Bug fixes:
- Fix reversed logic whether to enable changing password of a keystore entry
  (#2831542).
- Fix ClassCastException when displaying directory general names with Bouncy
  Castle >= 1.43.
- Fix display of CRLs that have no next update date.

Miscellaneous:
- Bundled Bouncy Castle provider jar upgraded to version 1.44; 1.43 is the
  new minimum required version.

New in Portecle 1.4 (2009-02-22):
---------------------------------

New features:
- Support for importing key pairs from OpenSSL PEM bundles.
- Support for encrypting private key in PEM export.
- Recognition and decoding support for some additional certificate
  extensions.
- Experimental: support for renewing self signed certificates.

Enhancements:
- Significant user interface improvements, including among others use of
  Java Desktop API to launch browsers and mail clients, more and improved
  default values in forms, and improved extension viewer dialog with hyperlink
  support for other resources in them (CRLs, certificates etc).
- Support changing look and feel without restarting the app.
- Support online help not bundled in jar (set helpbaseurl at build time).

Bug fixes:
- Fix problem with examining certificates from OpenSSL PEM bundles in
  certain security provider configurations.

Miscellaneous:
- Java SE 6 or later is now required.
- Bundled Bouncy Castle provider jar upgraded to 1.41.
- Removed built in support code for additional external look and feels;
  users who want them can have a look at Java's built in facilities for
  installing them.


src/main/net/sf/portecle/crypto/KeyStoreUtil.java
/*
 * KeyStoreUtil.java
 * This file is part of Portecle, a multipurpose keystore and certificate tool.
 *
 * Copyright © 2004 Wayne Grant, waynedgrant@hotmail.com
 *             2004-2017 Ville Skyttä, ville.skytta@iki.fi
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301, USA.
 */

package net.sf.portecle.crypto;

import static net.sf.portecle.FPortecle.RB;

import java.io.ByteArrayInputStream;
import java.io.File;
import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.io.FileOutputStream;
import java.io.IOException;
import java.security.GeneralSecurityException;
import java.security.KeyPair;
import java.security.KeyStore;
import java.security.KeyStoreException;
import java.security.NoSuchProviderException;
import java.security.Security;
import java.security.cert.Certificate;
import java.security.cert.CertificateException;
import java.security.cert.CertificateFactory;
import java.security.cert.X509Certificate;
import java.text.MessageFormat;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.Iterator;
import java.util.LinkedHashSet;

import org.bouncycastle.cert.X509CertificateHolder;
import org.bouncycastle.openssl.PEMDecryptorProvider;
import org.bouncycastle.openssl.PEMEncryptedKeyPair;
import org.bouncycastle.openssl.PEMKeyPair;
import org.bouncycastle.openssl.PEMParser;
import org.bouncycastle.openssl.PasswordFinder;
import org.bouncycastle.openssl.jcajce.JcaPEMKeyConverter;
import org.bouncycastle.openssl.jcajce.JcePEMDecryptorProviderBuilder;

/**
 * Provides utility methods for loading/saving keystores. The Bouncy Castle provider must be registered before using
 * this class to create or load BKS, UBER, or BCFKS type keystores.
 */
@SuppressWarnings("deprecation")
public final class KeyStoreUtil
{
	/**
	 * Dummy password to use for keystore entries in various contexts of keystores that do not support entry passwords.
	 */
	public static final char[] DUMMY_PASSWORD = "password".toCharArray();

	/** Map of available keystore types */
	private static final HashMap<KeyStoreType, Boolean> AVAILABLE_TYPES = new HashMap<>();

	/**
	 * Private to prevent construction.
	 */
	private KeyStoreUtil()
	{
		// Nothing to do
	}

	/**
	 * Gets the preferred (by us) KeyStore instance for the given keystore type.
	 *
	 * @param keyStoreType The keystore type
	 * @return The keystore
	 * @throws KeyStoreException No implementation found
	 */
	private static KeyStore getKeyStoreImpl(KeyStoreType keyStoreType)
	    throws KeyStoreException
	{
		KeyStore keyStore = null;
		if (keyStoreType == KeyStoreType.PKCS12)
		{
			// Prefer BC for PKCS #12 for now; the BC and SunJSSE 1.5+ implementations are incompatible in how
			// they handle empty/missing passwords; BC works consistently with char[0] on load and store (does
			// not accept nulls), SunJSSE throws division by zero with char[0] on load and store, works with
			// null on load, does not work with null on store.
			// Checked with BC 1.{29,40}, SunJSSE 1.5.0_0{3,4,14}, 1.6.0 (OpenJDK)
			try
			{
				keyStore = KeyStore.getInstance(keyStoreType.getTypeName(), "BC");
			}
			catch (NoSuchProviderException ex)
			{
				// Fall through
			}
		}
		if (keyStore == null)
		{
			try
			{
				keyStore = KeyStore.getInstance(keyStoreType.getTypeName());
			}
			catch (KeyStoreException e)
			{
				AVAILABLE_TYPES.put(keyStoreType, Boolean.FALSE);
				throw e;
			}
		}
		AVAILABLE_TYPES.put(keyStoreType, Boolean.TRUE);
		return keyStore;
	}

	/**
	 * Create a new, empty keystore.
	 *
	 * @param keyStoreType The keystore type to create
	 * @return The keystore
	 * @throws CryptoException Problem encountered creating the keystore
	 * @throws IOException An I/O error occurred
	 */
	public static KeyStore createKeyStore(KeyStoreType keyStoreType)
	    throws CryptoException, IOException
	{
		KeyStore keyStore;
		try
		{
			keyStore = getKeyStoreImpl(keyStoreType);
			keyStore.load(null, null);
		}
		catch (GeneralSecurityException ex)
		{
			throw new CryptoException(
			    MessageFormat.format(RB.getString("NoCreateKeystore.exception.message"), keyStoreType), ex);
		}
		return keyStore;
	}

	/**
	 * Load keystore entries from PEM reader into a new PKCS #12 keystore. The reader is not closed.
	 *
	 * @param reader reader to read entries from
	 * @param pwFinder object to get passwords from on demand
	 * @return new PKCS #12 keystore containing read entries, possibly empty
	 * @throws CertificateException
	 * @throws CryptoException Problem encountered creating the keystore
	 * @throws IOException An I/O error occurred
	 */
	public static KeyStore loadEntries(PEMParser reader, PasswordFinder pwFinder)
	    throws CertificateException, CryptoException, IOException
	{
		LinkedHashSet<KeyPair> keyPairs = new LinkedHashSet<>();
		LinkedHashSet<Certificate> certs = new LinkedHashSet<>();
		KeyStore keyStore = createKeyStore(KeyStoreType.PKCS12);

		CertificateFactory cf = CertificateFactory.getInstance(X509CertUtil.X509_CERT_TYPE);
		JcaPEMKeyConverter keyConverter = new JcaPEMKeyConverter();

		Object obj;
		while ((obj = reader.readObject()) != null)
		{
			if (obj instanceof PEMEncryptedKeyPair)
			{
				PEMDecryptorProvider decryptor = new JcePEMDecryptorProviderBuilder().build(pwFinder.getPassword());
				obj = ((PEMEncryptedKeyPair) obj).decryptKeyPair(decryptor);
			}
			if (obj instanceof PEMKeyPair)
			{
				keyPairs.add(keyConverter.getKeyPair((PEMKeyPair) obj));
			}
			else if (obj instanceof X509CertificateHolder)
			{
				ByteArrayInputStream bais = new ByteArrayInputStream(((X509CertificateHolder) obj).getEncoded());
				certs.add(cf.generateCertificate(bais));
			}
		}

		// Add key pairs
		for (KeyPair keyPair : keyPairs)
		{
			Certificate keyPairCert = null;
			for (Iterator<Certificate> it = certs.iterator(); it.hasNext();)
			{
				Certificate cert = it.next();
				if (cert.getPublicKey().equals(keyPair.getPublic()))
				{
					keyPairCert = cert;
					it.remove();
					break;
				}
			}

			if (keyPairCert != null)
			{
				String alias = "keypair";
				if (keyPairCert instanceof X509Certificate)
				{
					alias = X509CertUtil.getCertificateAlias((X509Certificate) keyPairCert);
				}

				KeyStore.PrivateKeyEntry entry =
				    new KeyStore.PrivateKeyEntry(keyPair.getPrivate(), new Certificate[] { keyPairCert });
				KeyStore.PasswordProtection prot = new KeyStore.PasswordProtection(DUMMY_PASSWORD);

				try
				{
					alias = findUnusedAlias(keyStore, alias);
					keyStore.setEntry(alias, entry, prot);
				}
				catch (KeyStoreException e)
				{
					throw new CryptoException(e);
				}
			}
		}

		// Add remaining certificates as trusted certificate entries
		for (Certificate cert : certs)
		{
			String alias = "certificate";
			if (cert instanceof X509Certificate)
			{
				alias = X509CertUtil.getCertificateAlias((X509Certificate) cert);
			}

			KeyStore.TrustedCertificateEntry entry = new KeyStore.TrustedCertificateEntry(cert);
			try
			{
				keyStore.setEntry(alias, entry, null);
			}
			catch (KeyStoreException e)
			{
				throw new CryptoException(e);
			}
		}

		return keyStore;
	}

	/**
	 * Check if a keystore type is available.
	 *
	 * @param keyStoreType the keystore type
	 * @return true if the keystore type is available, false otherwise
	 */
	public static boolean isAvailable(KeyStoreType keyStoreType)
	{
		Boolean available;
		if ((available = AVAILABLE_TYPES.get(keyStoreType)) != null)
		{
			return available;
		}
		try
		{
			// Populate AVAILABLE_TYPES
			getKeyStoreImpl(keyStoreType);
		}
		catch (KeyStoreException e)
		{
			// Ignore
		}
		return AVAILABLE_TYPES.get(keyStoreType);
	}

	/**
	 * Get available keystore types.
	 *
	 * @return available keystore types
	 */
	public static KeyStoreType[] getAvailableTypes()
	{
		// TODO: populate only once
		KeyStoreType[] known = KeyStoreType.values();
		ArrayList<KeyStoreType> available = new ArrayList<>();
		for (KeyStoreType type : known)
		{
			if (isAvailable(type))
			{
				available.add(type);
			}
		}
		return available.toArray(new KeyStoreType[available.size()]);
	}

	/**
	 * Load a Keystore from a file accessed by a password.
	 *
	 * @param keyStoreType The type of the keystore to open
	 * @param fKeyStore File to load keystore from
	 * @param cPassword Password of the keystore
	 * @return The keystore
	 * @throws CryptoException Problem encountered loading the keystore
	 * @throws FileNotFoundException If the keystore file does not exist, is a directory rather than a regular file, or
	 *             for some other reason cannot be opened for reading
	 */
	public static KeyStore loadKeyStore(File fKeyStore, char[] cPassword, KeyStoreType keyStoreType)
	    throws CryptoException, FileNotFoundException
	{
		KeyStore keyStore;
		try
		{
			keyStore = getKeyStoreImpl(keyStoreType);
		}
		catch (KeyStoreException ex)
		{
			throw new CryptoException(
			    MessageFormat.format(RB.getString("NoCreateKeystore.exception.message"), keyStoreType), ex);
		}

		try (FileInputStream fis = new FileInputStream(fKeyStore))
		{
			keyStore.load(fis, cPassword);
		}
		catch (FileNotFoundException ex)
		{
			throw ex;
		}
		// BC 1.58 workarounds for https://www.bouncycastle.org/jira/browse/BJA-691
		// * ClassCastException: trying to open PKCS #12 as BCFKS
		// * IllegalArgumentException: trying to open BCFKS as PKCS #12
		catch (GeneralSecurityException | IOException | ClassCastException | IllegalArgumentException ex)
		{
			throw new CryptoException(
			    MessageFormat.format(RB.getString("NoLoadKeystore.exception.message"), keyStoreType), ex);
		}

		return keyStore;
	}

	/**
	 * Load a PKCS #11 keystore accessed by a password.
	 *
	 * @param sPkcs11Provider The name of the PKCS #11 provider
	 * @param cPassword Password of the keystore
	 * @return The keystore
	 * @throws CryptoException Problem encountered loading the keystore
	 */
	public static KeyStore loadKeyStore(String sPkcs11Provider, char[] cPassword)
	    throws CryptoException
	{
		KeyStore keyStore;

		try
		{
			if (Security.getProvider(sPkcs11Provider) == null)
			{
				throw new CryptoException(
				    MessageFormat.format(RB.getString("NoSuchProvider.exception.message"), sPkcs11Provider));
			}
			keyStore = KeyStore.getInstance(KeyStoreType.PKCS11.name(), sPkcs11Provider);
		}
		catch (GeneralSecurityException ex)
		{
			throw new CryptoException(
			    MessageFormat.format(RB.getString("NoCreateKeystore.exception.message"), KeyStoreType.PKCS11), ex);
		}

		try
		{
			keyStore.load(null, cPassword);
		}
		catch (Exception ex)
		{
			throw new CryptoException(
			    MessageFormat.format(RB.getString("NoLoadKeystore.exception.message"), KeyStoreType.PKCS11), ex);
		}

		return keyStore;
	}

	/**
	 * Save a keystore to a file protected by a password.
	 *
	 * @param keyStore The keystore
	 * @param fKeyStoreFile The file to save the keystore to
	 * @param cPassword The password to protect the keystore with
	 * @return the saved keystore ready for further use
	 * @throws CryptoException Problem encountered saving the keystore
	 * @throws FileNotFoundException If the keystore file exists but is a directory rather than a regular file, does not
	 *             exist but cannot be created, or cannot be opened for any other reason
	 * @throws IOException An I/O error occurred
	 */
	public static KeyStore saveKeyStore(KeyStore keyStore, File fKeyStoreFile, char[] cPassword)
	    throws CryptoException, IOException
	{
		try (FileOutputStream fos = new FileOutputStream(fKeyStoreFile))
		{
			keyStore.store(fos, cPassword);
		}
		catch (GeneralSecurityException | IOException ex)
		{
			throw new CryptoException(RB.getString("NoSaveKeystore.exception.message"), ex);
		}

		// As of GNU classpath 0.92, we need to reload GKR keystores after storing them, otherwise
		// "masked envelope" IllegalStateExceptions occur when trying to access things in the stored keystore
		// again.
		if (KeyStoreType.valueOfType(keyStore.getType()) == KeyStoreType.GKR)
		{
			keyStore = loadKeyStore(fKeyStoreFile, cPassword, KeyStoreType.GKR);
		}

		return keyStore;
	}

	/**
	 * Find an unused alias in the keystore based on the given alias.
	 *
	 * @param keyStore the keystore
	 * @param alias the alias
	 * @return alias that is not in use in the keystore
	 * @throws KeyStoreException
	 */
	private static String findUnusedAlias(KeyStore keyStore, String alias)
	    throws KeyStoreException
	{
		if (keyStore.containsAlias(alias))
		{
			int i = 1;
			while (true)
			{
				String nextAlias = alias + " (" + i + ")";
				if (!keyStore.containsAlias(nextAlias))
				{
					alias = nextAlias;
					break;
				}
			}
		}
		return alias;
	}
}
